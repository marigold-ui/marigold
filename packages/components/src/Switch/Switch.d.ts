import { AriaSwitchProps } from '@react-types/switch';
import { ComponentProps } from '@marigold/types';
import { ThemeExtensionsWithParts } from '@marigold/system';
export interface SwitchThemeExtension
  extends ThemeExtensionsWithParts<
    'Switch',
    ['container', 'label', 'track', 'thumb']
  > {}
export declare type CustomSwitchProps =
  | 'size'
  | 'value'
  | 'onBlur'
  | 'onChange'
  | 'onFocus'
  | 'onKeyDown'
  | 'onKeyUp';
export interface SwitchProps
  extends Omit<AriaSwitchProps, 'isSelected'>,
    Omit<ComponentProps<'input'>, CustomSwitchProps> {
  checked?: boolean;
  variant?: string;
  size?: string;
  width?: string;
}
export declare const Switch: ({
  variant,
  size,
  width,
  checked,
  disabled,
  readOnly,
  defaultChecked,
  ...rest
}: SwitchProps) => JSX.Element;
//# sourceMappingURL=Switch.d.ts.map
