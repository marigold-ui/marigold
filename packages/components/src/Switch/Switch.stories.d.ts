import type { Meta, ComponentStory } from '@storybook/react';
import { Switch } from './Switch';
declare const _default: Meta<import('@storybook/react').Args>;
export default _default;
export declare const Basic: ComponentStory<typeof Switch>;
//# sourceMappingURL=Switch.stories.d.ts.map
