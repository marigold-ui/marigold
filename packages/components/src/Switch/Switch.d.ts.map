{"version":3,"file":"Switch.d.ts","sourceRoot":"","sources":["Switch.tsx"],"names":[],"mappings":"AAIA,OAAO,EAAE,eAAe,EAAE,MAAM,qBAAqB,CAAC;AAEtD,OAAO,EAAE,cAAc,EAAE,MAAM,iBAAiB,CAAC;AACjD,OAAO,EACL,wBAAwB,EAGzB,MAAM,kBAAkB,CAAC;AAM1B,MAAM,WAAW,oBACf,SAAQ,wBAAwB,CAC9B,QAAQ,EACR;IAAC,WAAW;IAAE,OAAO;IAAE,OAAO;IAAE,OAAO;CAAC,CACzC;CAAG;AAIN,oBAAY,iBAAiB,GACzB,MAAM,GACN,OAAO,GACP,QAAQ,GACR,UAAU,GACV,SAAS,GACT,WAAW,GACX,SAAS,CAAC;AAEd,MAAM,WAAW,WACf,SAAQ,IAAI,CAAC,eAAe,EAAE,YAAY,CAAC,EACzC,IAAI,CAAC,cAAc,CAAC,OAAO,CAAC,EAAE,iBAAiB,CAAC;IAClD,OAAO,CAAC,EAAE,OAAO,CAAC;IAClB,OAAO,CAAC,EAAE,MAAM,CAAC;IACjB,IAAI,CAAC,EAAE,MAAM,CAAC;IACd,KAAK,CAAC,EAAE,MAAM,CAAC;CAChB;AAGD,eAAO,MAAM,MAAM,mFAShB,WAAW,gBA+Fb,CAAC"}