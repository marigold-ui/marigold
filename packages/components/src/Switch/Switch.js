import React, { useRef } from 'react';
import { useFocusRing } from '@react-aria/focus';
import { useSwitch } from '@react-aria/switch';
import { useToggleState } from '@react-stately/toggle';
import { useComponentStyles, useStateProps } from '@marigold/system';
import { Box } from '../Box';
// Component
// ---------------
export const Switch = ({
  variant,
  size,
  width = '100%',
  checked,
  disabled,
  readOnly,
  defaultChecked,
  ...rest
}) => {
  const ref = useRef(null);
  // Adjust props to the react-aria API
  const props = {
    isSelected: checked,
    isDisabled: disabled,
    isReadOnly: readOnly,
    defaultSelected: defaultChecked,
    ...rest,
  };
  const state = useToggleState(props);
  const { inputProps } = useSwitch(props, state, ref);
  const { isFocusVisible, focusProps } = useFocusRing();
  const stateProps = useStateProps({
    checked: state.isSelected,
    disabled: disabled,
    readOnly: readOnly,
    focus: isFocusVisible,
  });
  const styles = useComponentStyles(
    'Switch',
    { variant, size },
    { parts: ['container', 'label', 'track', 'thumb'] }
  );
  return React.createElement(
    Box,
    {
      as: 'label',
      __baseCSS: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-between',
        gap: '1ch',
        position: 'relative',
        width,
      },
      css: styles.container,
    },
    React.createElement(Box, {
      as: 'input',
      ref: ref,
      css: {
        position: 'absolute',
        width: '100%',
        height: '100%',
        top: 0,
        left: 0,
        zIndex: 1,
        opacity: 0.0001,
        cursor: inputProps.disabled ? 'not-allowed' : 'pointer',
      },
      ...inputProps,
      ...focusProps,
    }),
    props.children &&
      React.createElement(Box, { css: styles.label }, props.children),
    React.createElement(
      Box,
      {
        __baseCSS: {
          position: 'relative',
          width: 48,
          height: 24,
          bg: '#dee2e6',
          borderRadius: 20,
        },
        css: styles.track,
        ...stateProps,
      },
      React.createElement(Box, {
        __baseCSS: {
          display: 'block',
          position: 'absolute',
          top: 1,
          left: 0,
          willChange: 'transform',
          transform: 'translateX(1px)',
          transition: 'all 0.1s cubic-bezier(.7, 0, .3, 1)',
          height: 22,
          width: 22,
          borderRadius: 9999,
          bg: '#fff',
          '&:checked': {
            transform: 'translateX(calc(47px - 100%))',
          },
        },
        css: styles.thumb,
        ...stateProps,
      })
    )
  );
};
//# sourceMappingURL=Switch.js.map
