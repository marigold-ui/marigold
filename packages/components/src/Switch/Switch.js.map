{"version":3,"file":"Switch.js","sourceRoot":"","sources":["Switch.tsx"],"names":[],"mappings":"AAAA,OAAO,KAAK,EAAE,EAAE,MAAM,EAAE,MAAM,OAAO,CAAC;AACtC,OAAO,EAAE,YAAY,EAAE,MAAM,mBAAmB,CAAC;AACjD,OAAO,EAAE,SAAS,EAAE,MAAM,oBAAoB,CAAC;AAC/C,OAAO,EAAE,cAAc,EAAE,MAAM,uBAAuB,CAAC;AAIvD,OAAO,EAEL,kBAAkB,EAClB,aAAa,GACd,MAAM,kBAAkB,CAAC;AAE1B,OAAO,EAAE,GAAG,EAAE,MAAM,QAAQ,CAAC;AA6B7B,YAAY;AACZ,kBAAkB;AAClB,MAAM,CAAC,MAAM,MAAM,GAAG,CAAC,EACrB,OAAO,EACP,IAAI,EACJ,KAAK,GAAG,MAAM,EACd,OAAO,EACP,QAAQ,EACR,QAAQ,EACR,cAAc,EACd,GAAG,IAAI,EACK,EAAE,EAAE;IAChB,MAAM,GAAG,GAAG,MAAM,CAAmB,IAAI,CAAC,CAAC;IAC3C,qCAAqC;IACrC,MAAM,KAAK,GAAG;QACZ,UAAU,EAAE,OAAO;QACnB,UAAU,EAAE,QAAQ;QACpB,UAAU,EAAE,QAAQ;QACpB,eAAe,EAAE,cAAc;QAC/B,GAAG,IAAI;KACR,CAAC;IAEF,MAAM,KAAK,GAAG,cAAc,CAAC,KAAK,CAAC,CAAC;IACpC,MAAM,EAAE,UAAU,EAAE,GAAG,SAAS,CAAC,KAAK,EAAE,KAAK,EAAE,GAAG,CAAC,CAAC;IACpD,MAAM,EAAE,cAAc,EAAE,UAAU,EAAE,GAAG,YAAY,EAAE,CAAC;IACtD,MAAM,UAAU,GAAG,aAAa,CAAC;QAC/B,OAAO,EAAE,KAAK,CAAC,UAAU;QACzB,QAAQ,EAAE,QAAQ;QAClB,QAAQ,EAAE,QAAQ;QAClB,KAAK,EAAE,cAAc;KACtB,CAAC,CAAC;IAEH,MAAM,MAAM,GAAG,kBAAkB,CAC/B,QAAQ,EACR,EAAE,OAAO,EAAE,IAAI,EAAE,EACjB,EAAE,KAAK,EAAE,CAAC,WAAW,EAAE,OAAO,EAAE,OAAO,EAAE,OAAO,CAAC,EAAE,CACpD,CAAC;IAEF,OAAO,CACL,oBAAC,GAAG,IACF,EAAE,EAAC,OAAO,EACV,SAAS,EAAE;YACT,OAAO,EAAE,MAAM;YACf,UAAU,EAAE,QAAQ;YACpB,cAAc,EAAE,eAAe;YAC/B,GAAG,EAAE,KAAK;YACV,QAAQ,EAAE,UAAU;YACpB,KAAK;SACN,EACD,GAAG,EAAE,MAAM,CAAC,SAAS;QAErB,oBAAC,GAAG,IACF,EAAE,EAAC,OAAO,EACV,GAAG,EAAE,GAAG,EACR,GAAG,EAAE;gBACH,QAAQ,EAAE,UAAU;gBACpB,KAAK,EAAE,MAAM;gBACb,MAAM,EAAE,MAAM;gBACd,GAAG,EAAE,CAAC;gBACN,IAAI,EAAE,CAAC;gBACP,MAAM,EAAE,CAAC;gBACT,OAAO,EAAE,MAAM;gBACf,MAAM,EAAE,UAAU,CAAC,QAAQ,CAAC,CAAC,CAAC,aAAa,CAAC,CAAC,CAAC,SAAS;aACxD,KACG,UAAU,KACV,UAAU,GACd;QACD,KAAK,CAAC,QAAQ,IAAI,oBAAC,GAAG,IAAC,GAAG,EAAE,MAAM,CAAC,KAAK,IAAG,KAAK,CAAC,QAAQ,CAAO;QACjE,oBAAC,GAAG,IACF,SAAS,EAAE;gBACT,QAAQ,EAAE,UAAU;gBACpB,KAAK,EAAE,EAAE;gBACT,MAAM,EAAE,EAAE;gBACV,EAAE,EAAE,SAAS;gBACb,YAAY,EAAE,EAAE;aACjB,EACD,GAAG,EAAE,MAAM,CAAC,KAAK,KACb,UAAU;YAEd,oBAAC,GAAG,IACF,SAAS,EAAE;oBACT,OAAO,EAAE,OAAO;oBAChB,QAAQ,EAAE,UAAU;oBACpB,GAAG,EAAE,CAAC;oBACN,IAAI,EAAE,CAAC;oBAEP,UAAU,EAAE,WAAW;oBACvB,SAAS,EAAE,iBAAiB;oBAC5B,UAAU,EAAE,qCAAqC;oBAEjD,MAAM,EAAE,EAAE;oBACV,KAAK,EAAE,EAAE;oBAET,YAAY,EAAE,IAAI;oBAClB,EAAE,EAAE,MAAM;oBAEV,WAAW,EAAE;wBACX,SAAS,EAAE,+BAA+B;qBAC3C;iBACF,EACD,GAAG,EAAE,MAAM,CAAC,KAAK,KACb,UAAU,GACd,CACE,CACF,CACP,CAAC;AACJ,CAAC,CAAC"}