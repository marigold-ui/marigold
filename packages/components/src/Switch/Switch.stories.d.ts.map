{"version":3,"file":"Switch.stories.d.ts","sourceRoot":"","sources":["Switch.stories.tsx"],"names":[],"mappings":"AACA,OAAO,KAAK,EAAE,IAAI,EAAE,cAAc,EAAE,MAAM,kBAAkB,CAAC;AAC7D,OAAO,EAAE,MAAM,EAAE,MAAM,UAAU,CAAC;;AAElC,wBAoCU;AAEV,eAAO,MAAM,KAAK,EAAE,cAAc,CAAC,OAAO,MAAM,CAE/C,CAAC"}