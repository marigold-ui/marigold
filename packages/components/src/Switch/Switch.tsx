import React, { forwardRef } from 'react';
import { useFocusRing } from '@react-aria/focus';
import { useSwitch } from '@react-aria/switch';
import { useObjectRef } from '@react-aria/utils';
import { useToggleState } from '@react-stately/toggle';
import { AriaSwitchProps } from '@react-types/switch';

import { HtmlProps } from '@marigold/types';
import { cn, createVar, useStateProps, useClassNames } from '@marigold/system';

// Theme Extension
// ---------------

// Props
// ---------------
export type CustomSwitchProps =
  | 'size'
  | 'value'
  | 'onBlur'
  | 'onChange'
  | 'onFocus'
  | 'onKeyDown'
  | 'onKeyUp';

export interface SwitchProps
  extends Omit<AriaSwitchProps, 'isSelected'>,
    Omit<HtmlProps<'input'>, CustomSwitchProps> {
  checked?: boolean;
  variant?: string;
  size?: string;
  width?: string;
}

// Component
// ---------------
export const Switch = forwardRef<HTMLInputElement, SwitchProps>(
  (
    {
      variant,
      size,
      width = '100%',
      checked,
      disabled,
      readOnly,
      defaultChecked,
      ...rest
    },
    ref
  ) => {
    const inputRef = useObjectRef(ref);
    // Adjust props to the react-aria API
    const props = {
      isSelected: checked,
      isDisabled: disabled,
      isReadOnly: readOnly,
      defaultSelected: defaultChecked,
      ...rest,
    };

    const state = useToggleState(props);
    const { inputProps } = useSwitch(props, state, inputRef);
    const { isFocusVisible, focusProps } = useFocusRing();
    const stateProps = useStateProps({
      selected: state.isSelected,
      disabled: disabled,
      readOnly: readOnly,
      focus: isFocusVisible,
    });

    const classNames = useClassNames({ component: 'Switch', size, variant });
    return (
      <label
        className={cn(
          'group/switch',
          'relative flex items-center justify-between gap-[1ch]'
        )}
        style={createVar({ fieldWidth: width })}
        {...stateProps}
      >
        <input
          ref={inputRef}
          className="absolute left-0 top-0 z-[1] h-full w-full opacity-[0.0001] disabled:cursor-not-allowed"
          {...inputProps}
          {...focusProps}
        />
        {props.children && (
          <div
            className={classNames.label}
            // css={styles.label}
          >
            {props.children}
          </div>
        )}
        <div
          className={cn(
            'relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl',
            classNames.track
          )}
        >
          <div
            className={cn(
              'transition-[all 0.1s cubic-bezier(.7, 0, .3, 1)] checked:translate-x-[calc(47px - 100%)] absolute left-0 top-0 block translate-x-[1px] rounded-full will-change-transform',
              'group-selected/switch:translate-x-[calc(47px_-_100%)]',
              classNames.thumb
            )}
            {...stateProps}
          />
        </div>
      </label>
    );
  }
);
