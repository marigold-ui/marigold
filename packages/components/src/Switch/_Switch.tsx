import { forwardRef } from 'react';

import { useFocusRing } from '@react-aria/focus';
import { useSwitch } from '@react-aria/switch';
import { useObjectRef } from '@react-aria/utils';

import { useToggleState } from '@react-stately/toggle';

import { AriaSwitchProps } from '@react-types/switch';

import { cn, createVar, useClassNames, useStateProps } from '@marigold/system';
import { HtmlProps } from '@marigold/types';

// Theme Extension
// ---------------

// Props
// ---------------
export type CustomSwitchProps =
  | 'size'
  | 'value'
  | 'onBlur'
  | 'onChange'
  | 'onFocus'
  | 'onKeyDown'
  | 'onKeyUp';

export interface SwitchProps
  extends Omit<AriaSwitchProps, 'isSelected'>,
    Omit<HtmlProps<'input'>, CustomSwitchProps | 'children' | 'className'> {
  selected?: boolean;
  variant?: string;
  size?: string;
  width?: string;
}

// Component
// ---------------
export const Switch = forwardRef<HTMLInputElement, SwitchProps>(
  (
    {
      variant,
      size,
      width = '100%',
      selected,
      disabled,
      readOnly,
      defaultChecked,
      ...rest
    },
    ref
  ) => {
    const inputRef = useObjectRef(ref);
    // Adjust props to the react-aria API
    const props = {
      isSelected: selected,
      isDisabled: disabled,
      isReadOnly: readOnly,
      defaultSelected: defaultChecked,
      ...rest,
    };

    const state = useToggleState(props);
    const { inputProps } = useSwitch(props, state, inputRef);
    const { isFocusVisible, focusProps } = useFocusRing();
    const stateProps = useStateProps({
      selected: state.isSelected,
      disabled: disabled,
      readOnly: readOnly,
      focus: isFocusVisible,
    });
    const classNames = useClassNames({ component: 'Switch', size, variant });
    return (
      <label
        className={cn(
          'group/switch',
          'w-[var(--switchWidth)]',
          'relative flex items-center justify-between gap-[1ch]'
        )}
        style={createVar({ switchWidth: width })}
        {...stateProps}
      >
        <input
          ref={inputRef}
          className="absolute left-0 top-0 z-[1] h-full w-full cursor-pointer opacity-[0.0001] disabled:cursor-not-allowed"
          {...inputProps}
          {...focusProps}
        />
        {props.children && <>{props.children}</>}
        <div
          className={cn(
            'relative h-6 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl',
            classNames.track
          )}
        >
          <div
            className={cn(
              'h-[22px] w-[22px]',
              'cubic-bezier(.7,0,.3,1)',
              'absolute left-0 top-px',
              'block translate-x-[1px] rounded-full transition-all duration-100 ease-in-out will-change-transform',
              'group-selected/switch:translate-x-[calc(47px_-_100%)]',
              classNames.thumb
            )}
          />
        </div>
      </label>
    );
  }
);
