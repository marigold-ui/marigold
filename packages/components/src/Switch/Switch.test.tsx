import React from 'react';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { Theme } from '@marigold/system';
import { Switch } from './Switch';
import { cva } from 'class-variance-authority';
import userEvent from '@testing-library/user-event';
import { setup } from '../test.utils';

const theme: Theme = {
  name: 'switch test',
  components: {
    Switch: {
      container: cva(),
      track: cva(
        [
          'border-input-border border bg-white shadow-[inset_0px_0px_1px',
          'checked:bg-switch-track-primary checked:shadow-switch-track-checked',
          'disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-offset[3]',
          'focus:outline-switch-track-outline-focus',
        ],
        {
          variants: {
            size: {
              large: 'w-[96] h-[48] rounded-[40]',
            },
          },
        }
      ),
      thumb: cva(
        ['bg-root-body', 'h-4 w-4'],

        // 'shadow-[inset_1px_1px_4px] shadow-switch-thumb-shadow disabled:bg-switch-thumb-disabled',
        {
          variants: {
            size: {
              large:
                'top-2 w-[44] h-[44] checked:translate-x-[calc(95px_-_44px)]',
            },
          },
        }
      ),
      label: cva(''),
    },
  },
};

const user = userEvent.setup();
const getSwitchParts = () => {
  const label: HTMLLabelElement = screen.getByText('Label');
  // eslint-disable-next-line testing-library/no-node-access
  const container: HTMLElement = label.parentElement!;
  // eslint-disable-next-line testing-library/no-node-access
  const track = container.lastChild! as HTMLElement;
  // eslint-disable-next-line testing-library/no-node-access
  const thumb = track.lastChild! as HTMLElement;

  const input: HTMLInputElement = screen.getByRole('switch');

  return { label, input, container, track, thumb };
};

const { render } = setup({ theme });

test('supports base styling', () => {
  render(<Switch>Label</Switch>);
  const { label, container, track, thumb } = getSwitchParts();

  expect(label.className).toMatchInlineSnapshot(`""`);
  expect(container.className).toMatchInlineSnapshot(
    `"group/switch relative flex items-center justify-between gap-[1ch]"`
  );
  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
  expect(thumb.className).toMatchInlineSnapshot(
    `"transition-[all 0.1s cubic-bezier(.7, 0, .3, 1)] checked:translate-x-[calc(47px - 100%)] absolute left-0 top-0 block translate-x-[1px] rounded-full will-change-transform group-selected/switch:translate-x-[calc(47px_-_100%)] bg-root-body h-4 w-4"`
  );
});

test('supports a custom variant', () => {
  render(<Switch variant="custom">Label</Switch>);
  const { container, track, thumb } = getSwitchParts();

  expect(container.className).toMatchSnapshot();
  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
  expect(thumb.className).toMatchInlineSnapshot(
    `"transition-[all 0.1s cubic-bezier(.7, 0, .3, 1)] checked:translate-x-[calc(47px - 100%)] absolute left-0 top-0 block translate-x-[1px] rounded-full will-change-transform group-selected/switch:translate-x-[calc(47px_-_100%)] bg-root-body h-4 w-4"`
  );
});

test('supports a size', () => {
  render(<Switch size="medium">Label</Switch>);
  const { track } = getSwitchParts();

  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
});

test('takes full width by default', () => {
  render(<Switch>Label</Switch>);
  const { container } = getSwitchParts();
  expect(container.className).toMatchInlineSnapshot(
    `"group/switch relative flex items-center justify-between gap-[1ch]"`
  );
});

test('allows to set width via prop', () => {
  render(<Switch width="large">Label</Switch>);
  const { container } = getSwitchParts();
  expect(container.className).toMatchSnapshot();
});

test('supports disabled prop', () => {
  render(<Switch disabled>Label</Switch>);
  const { input, thumb, track } = getSwitchParts();

  expect(input).toBeDisabled();
  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
  expect(thumb.className).toMatchInlineSnapshot(
    `"transition-[all 0.1s cubic-bezier(.7, 0, .3, 1)] checked:translate-x-[calc(47px - 100%)] absolute left-0 top-0 block translate-x-[1px] rounded-full will-change-transform group-selected/switch:translate-x-[calc(47px_-_100%)] bg-root-body h-4 w-4"`
  );
});

test('renders hidden <input> element', () => {
  render(<Switch>Label</Switch>);
  const { input } = getSwitchParts();
  expect(input instanceof HTMLInputElement).toBeTruthy();
});

test('toggle switch per click', () => {
  render(<Switch>Label</Switch>);

  const { input, track } = getSwitchParts();

  fireEvent.click(input);
  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
  expect(input.checked).toBeTruthy();

  fireEvent.click(input);
  expect(track.className).toMatchInlineSnapshot(
    `"relative h-6 w-12 flex-shrink-0 flex-grow-0 basis-12 rounded-3xl border-input-border border bg-white shadow-[inset_0px_0px_1px checked:bg-switch-track-primary checked:shadow-switch-track-checked disabled:opacity-[0.5] disabled:shadow-switch-track-shadow disabled:bg-switch-track-shadow focus:outline-none focus:outline-switch-track-outline-focus"`
  );
  expect(input.checked).toBeFalsy();
});

test('focus element and toggle switch per keyboard space', async () => {
  render(<Switch>Label</Switch>);

  const { input, track } = getSwitchParts();
  user.tab();

  // eslint-disable-next-line testing-library/await-async-utils
  waitFor(() => expect(track).toHaveAttribute('data-focus'));
  user.keyboard('{space}');
  // eslint-disable-next-line testing-library/await-async-utils
  waitFor(() => expect(track).toHaveStyle(`background-color: orange`));
  // eslint-disable-next-line testing-library/await-async-utils
  waitFor(() => expect(input.checked).toBeTruthy());

  user.keyboard('{space}');
  // eslint-disable-next-line testing-library/await-async-utils
  waitFor(() => expect(track).toHaveStyle(`background-color: blue`));
  // eslint-disable-next-line testing-library/await-async-utils
  waitFor(() => expect(input.checked).toBeFalsy());
});

test('supports default checked', () => {
  render(<Switch defaultChecked>Label</Switch>);

  const { input } = getSwitchParts();

  expect(input.checked).toBeTruthy();
  fireEvent.click(input);
  expect(input.checked).toBeFalsy();
});

test('supports controlled component usage', () => {
  const onChange = jest.fn();
  render(<Switch onChange={onChange}>Label</Switch>);

  const { input } = getSwitchParts();

  fireEvent.click(input);
  expect(onChange).toHaveBeenCalledWith(true);
  expect(input.checked).toBeTruthy();

  fireEvent.click(input);
  expect(onChange).toHaveBeenCalledWith(false);
  expect(input.checked).toBeFalsy();
});

test('forwards ref', () => {
  const ref = React.createRef<HTMLInputElement>();
  render(<Switch ref={ref}>Label</Switch>);

  expect(ref.current).toBeInstanceOf(HTMLInputElement);
});
