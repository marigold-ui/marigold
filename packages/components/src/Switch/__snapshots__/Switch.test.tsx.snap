// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows to set width via prop 1`] = `"group/switch relative flex items-center justify-between gap-[1ch]"`;

exports[`supports a custom variant 1`] = `"group/switch relative flex items-center justify-between gap-[1ch]"`;
